File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverse(ListNode *h){
        ListNode *p=NULL,*c=h, *n;
        
        while(c){
            n=c->next;
            c->next=p;
            p=c;
            c=n;
        }
        
        return p;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
         if(head==NULL || head->next==NULL || k<2)return head;
        int count=0;
        ListNode *t=head,*p;
        while(t && count<k){
            p=t;
            t=t->next;
            count++;
        }
        if(count !=k)return head;
        p->next=NULL;
        ListNode *head1=reverse(head);
        ListNode *n=reverseKGroup(t,k);
        head->next=n;
        return head1;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public static ListNode reverse(ListNode head){
        ListNode curr = head;
        ListNode temp =  head;
        ListNode prev = null;

        while(curr!= null){
            temp = curr.next;
            curr.next = prev;
            prev  = curr;
            curr = temp;
        }

        return prev;
    }
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null || head.next == null) return head;

        int count = 0;
        ListNode curr = head;
        ListNode temp = head;
        while (count < k &&  curr != null){
            temp = curr;
            curr = curr.next;
            count ++;
        }
        if(count<k)return head;
        temp.next = null;
        ListNode start = reverse(head);  
        ListNode nextgroup = reverseKGroup(curr, k);
        head.next = nextgroup;
        head = start;

        return head;

    }
}